### PR TITLE
dbeaver/dbeaver#19437 open external files

### LINKED ISSUE #19347: Pictures from Value panel are incorrectly opened in External editor
### Description

Pictures from Value panel are incorrectly opened in External editor

### DBeaver Version

23.0.0

### Operating System

win10

### Database and driver

_No response_

### Steps to reproduce

1. Open some table with a blob column (containing pictures)
2. Open value panel - picture is shown correctly.
image

4. Open picture at external editor 
image

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Destrolaric**
Closes dbeaver/dbeaver#19347
Acceptance: 
Images opened in their proper format when `Open in external editor` pressed. (Actual only for browser viewer)

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/data/IStreamValueEditor.java [hunk 1]
```diff
@@ -67,4 +69,6 @@ void contributeSettings(@NotNull IContributionManager manager, @NotNull CONTROL
 
     void disposeEditor();
 
+    File getExternalFile();
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/data/IStreamValueEditor.java]:
```suggestion
    @Nullable
    default File getExternalFile(@NotNull CONTROL control) {
        return null;
    }
```

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/data/managers/ContentValueManager.java [hunk 2]
```diff
@@ -95,7 +97,20 @@ public void run() {
                             DBWorkbench.getPlatformUI().showError("Data is empty", "Can not save null data value");
                         }
                         if (value instanceof DBDContent) {
-                            getDBDContent(value);
+                            boolean isExternalFileOpened = false;
+                            if (activeEditor instanceof ContentPanelEditor) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/data/managers/ContentValueManager.java]: It's guaranteed to be `ContentPanelEditor` when opened as a panel.

File: plugins/org.jkiss.dbeaver.ui.editors.hex/src/org/jkiss/dbeaver/ui/data/BinaryPanelEditor.java [hunk 3]
```diff
@@ -159,4 +160,8 @@ public void disposeEditor() {
 
     }
 
+    @Override
+    public File getExternalFile() {
+        return null;
+    }
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.hex/src/org/jkiss/dbeaver/ui/data/BinaryPanelEditor.java]: No need to override it

File: plugins/org.jkiss.dbeaver.ui.editors.image/src/org/jkiss/dbeaver/ui/controls/imageview/AbstractImageViewer.java [hunk 4]
```diff
@@ -33,4 +35,8 @@ public AbstractImageViewer(Composite parent, int style) {
     public void fillToolBar(IContributionManager toolBar) {
 
     }
+
+    public File getExternalFile() {
+        return null;
+    }
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.image/src/org/jkiss/dbeaver/ui/controls/imageview/AbstractImageViewer.java]: No need to override it

File: plugins/org.jkiss.dbeaver.ui.editors.image/src/org/jkiss/dbeaver/ui/data/managers/image/ImagePanelEditor.java [hunk 5]
```diff
@@ -104,4 +112,10 @@ public void disposeEditor() {
 
     }
 
+    @Override
+    public File getExternalFile() {
+        return imageViewer.getExternalFile();
+    }
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.image/src/org/jkiss/dbeaver/ui/data/managers/image/ImagePanelEditor.java]: Use the `control` argument instead of storing a reference to the viewer. Editors are singletons. You can't do that.

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/data/editors/ContentPanelEditor.java [hunk 6]
```diff
@@ -307,6 +307,11 @@ public StreamValueManagerDescriptor getCurrentStreamManager() {
         return curStreamManager;
     }
 
+    @Nullable
+    public IStreamValueEditor<Control> getStreamEditor() {
+        return streamEditor;
+    }
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/data/editors/ContentPanelEditor.java:313]: See previous comment

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/data/managers/AbstractTextPanelEditor.java [hunk 7]
```diff
@@ -319,6 +320,11 @@ public void extractEditorValue(@NotNull DBRProgressMonitor monitor, @NotNull Sty
         }
     }
 
+    @Override
+    public File getExternalFile() {
+        return null;
+    }
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/data/managers/AbstractTextPanelEditor.java]: No need to override it

**Review by ShadelessFox — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/data/IStreamValueEditor.java [hunk 8]
```diff
@@ -67,4 +70,7 @@ void contributeSettings(@NotNull IContributionManager manager, @NotNull CONTROL
 
     void disposeEditor();
 
+    @Nullable
+    File getExternalFile(@NotNull CONTROL control);
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/data/IStreamValueEditor.java]: Let's move it to a separate interface (e.g. IStreamValueEditorPersistent) and implement it only in browser. 
And use nio Path instead of File.

Comment (serge-rider) [reply plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/data/IStreamValueEditor.java]: + javadocs

**Review by serge-rider — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/data/IStreamValueEditorPersistent.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+ */
+package org.jkiss.dbeaver.ui.data;
+
+import org.eclipse.swt.widgets.Control;
+import org.jkiss.code.NotNull;
+import org.jkiss.code.Nullable;
+
+import java.nio.file.Path;
+
+public interface IStreamValueEditorPersistent <CONTROL extends Control> {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/data/IStreamValueEditorPersistent.java]: Why doesn't it inherit from `IStreamValueEditor` to introduce constraints? Also leading space before `<`

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/data/IStreamValueEditorPersistent.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+    /**
+     * @param control Control to acquire filepath from
+     * @return filepath of external file or null if not found
+     */
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/data/IStreamValueEditorPersistent.java]: This is a bad Javadoc without a body.

The `control` is not used to acquire the file path. It can be obtained from anywhere. It does not strictly need to be the `control` which provides the file path. It's just convenience since the editor itself doesn't know its control. TL;DR: call it `editor control` or something like that

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/data/managers/ContentValueManager.java [hunk 11]
```diff
@@ -95,7 +99,20 @@ public void run() {
...
                         }
                         if (value instanceof DBDContent) {
-                            getDBDContent(value);
+                            boolean isExternalFileOpened = false;
+                            IStreamValueEditor<Control> streamEditor
+                                = ((ContentPanelEditor) activeEditor).getStreamEditor();
+                            if (streamEditor instanceof IStreamValueEditorPersistent) {
+                                Path externalFilePath =
+                                    ((IStreamValueEditorPersistent) streamEditor).getExternalFilePath(activeEditor.getControl());
+                                File externalFile = externalFilePath == null ? null : externalFilePath.toFile();
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/data/managers/ContentValueManager.java]: What's the point of converting back to `File`? It introduces redundant null check here

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/data/managers/ContentValueManager.java [hunk 12]
```diff
@@ -131,6 +148,19 @@ public void run() {
...
 
+    private static boolean openExternalFile(@NotNull File file) {
+        if (!file.exists()) {
+            return false;
+        }
+        try {
+            Desktop.getDesktop().open(file);
+            return true;
+        } catch (IOException e) {
+            log.error(e);
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/data/managers/ContentValueManager.java]: "Unable to open external file..."

File: plugins/org.jkiss.dbeaver.ui.editors.image/src/org/jkiss/dbeaver/ui/controls/imageview/AbstractImageViewer.java [hunk 13]
```diff
@@ -33,4 +35,8 @@ public AbstractImageViewer(Composite parent, int style) {
     public void fillToolBar(IContributionManager toolBar) {
 
     }
+
+    public Path getExternalFilePath() {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.image/src/org/jkiss/dbeaver/ui/controls/imageview/AbstractImageViewer.java:41]: Mind adding `@Nullable` annotation?

File: plugins/org.jkiss.dbeaver.ui.editors.image/src/org/jkiss/dbeaver/ui/controls/imageview/BrowserImageViewer.java [hunk 14]
```diff
@@ -189,4 +189,7 @@ private void clearTempFile() {
         tempFile = null;
     }
 
+    public Path getExternalFilePath() {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.image/src/org/jkiss/dbeaver/ui/controls/imageview/BrowserImageViewer.java:195]: Mind adding `@Nullable` and `@Override` annotations?

File: plugins/org.jkiss.dbeaver.ui.editors.image/src/org/jkiss/dbeaver/ui/data/managers/image/ImagePanelEditor.java [hunk 15]
```diff
@@ -32,15 +32,19 @@
...
 import java.io.InputStream;
+import java.nio.file.Path;
 
 /**
 * ImagePanelEditor
 */
-public class ImagePanelEditor implements IStreamValueEditor<AbstractImageViewer> {
+public class ImagePanelEditor implements IStreamValueEditor<AbstractImageViewer>, IStreamValueEditorPersistent<AbstractImageViewer> {
+
+    private AbstractImageViewer imageViewer;
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.image/src/org/jkiss/dbeaver/ui/data/managers/image/ImagePanelEditor.java]: Unrelated change

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/data/managers/AbstractTextPanelEditor.java [hunk 16]
```diff
@@ -319,6 +320,10 @@ public void extractEditorValue(@NotNull DBRProgressMonitor monitor, @NotNull Sty
         }
     }
 
+    public File getExternalFile(StyledText control) {
+        return null;
+    }
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/data/managers/AbstractTextPanelEditor.java]: Unrelated change

File: plugins/org.jkiss.dbeaver.ui.editors.image/src/org/jkiss/dbeaver/ui/data/managers/image/ImagePanelEditor.java [hunk 17]
```diff
@@ -32,15 +32,19 @@
...
 
 import java.io.IOException;
 import java.io.InputStream;
+import java.nio.file.Path;
 
 /**
 * ImagePanelEditor
 */
-public class ImagePanelEditor implements IStreamValueEditor<AbstractImageViewer> {
+public class ImagePanelEditor implements IStreamValueEditor<AbstractImageViewer>, IStreamValueEditorPersistent<AbstractImageViewer> {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.image/src/org/jkiss/dbeaver/ui/data/managers/image/ImagePanelEditor.java]: See comment in the `IStreamValueEditorPersistent.java`

```suggestion
public class ImagePanelEditor implements IStreamValueEditorPersistent<AbstractImageViewer> {
```

**Review by ShadelessFox — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.ui.editors.image/src/org/jkiss/dbeaver/ui/data/managers/image/ImagePanelEditor.java [hunk 18]
```diff
@@ -49,7 +50,10 @@ public AbstractImageViewer createControl(IValueController valueController) {
             .getContainer()
             .getPreferenceStore();
         if (preferenceStore.getBoolean(ResultSetPreferences.RESULT_IMAGE_USE_BROWSER_BASED_RENDERER)) {
-            return new BrowserImageViewer(valueController.getEditPlaceholder(), SWT.NONE);
+            return new BrowserImageViewer(
+                valueController.getEditPlaceholder(),
+                SWT.NONE
+            );
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.image/src/org/jkiss/dbeaver/ui/data/managers/image/ImagePanelEditor.java]: Redundant change

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/data/IStreamValueEditorPersistent.java [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * Value editor which provides external files
+ * @param <CONTROL> editor control
+ */
+public interface IStreamValueEditorPersistent<CONTROL extends Control> extends IStreamValueEditor<CONTROL> {
+
+    /**
+     * 
+     *
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/data/IStreamValueEditorPersistent.java]: Mind adding body?

**Review by ShadelessFox — APPROVED** [review]

**Review by serge-rider — APPROVED** [review]

**Review by LonwoLonwo — APPROVED** [review]

**Comment by uslss** [comment]
verified (SQLite, SQL Server)
